File: src/web-client/key_utils.py
import os
import json
import tkinter as tk
from tkinter import filedialog
from typing import Tuple, Optional
import base64
from nacl.public import PrivateKey
from nacl.bindings import (
    crypto_aead_xchacha20poly1305_ietf_KEYBYTES,
    crypto_aead_xchacha20poly1305_ietf_NPUBBYTES,
)
import nacl.utils
from nacl import pwhash
from encryption_helper import EncryptionHelper  
from constants import BINARY_EXTENSION, KEYS_PATH, MASTER_KEY_PATH
import os

def generate_sodium_keypair() -> tuple[str, str]:
    """
    Generate a public/private keypair using libsodium (PyNaCl) and return as base64-encoded strings.

    Returns:
        tuple[str, str]: (public_key_base64, private_key_base64)
    """
    private_key = PrivateKey.generate()
    public_key = private_key.public_key
    public_b64 = base64.b64encode(bytes(public_key)).decode('utf-8')
    private_b64 = base64.b64encode(bytes(private_key)).decode('utf-8')
    return public_b64, private_b64

def save_keys_to_json_file(public_key_b64: str, private_key_b64: str) -> Tuple[bool, Optional[str]]:
    """
    Save public and private keys (base64-encoded) to a JSON file using a file dialog.

    Args:
        public_key_b64 (str): Base64-encoded public key.
        private_key_b64 (str): Base64-encoded private key.

    Returns:
        Tuple[bool, Optional[str]]: A tuple containing:
            - bool: True if successful, False if user cancelled or error occurred
            - Optional[str]: The path where the file was saved, or None if failed
    """
    try:
        # Create and hide the root window
        root = tk.Tk()
        root.attributes('-topmost', True)  
        root.withdraw()  

        # Prepare the data to save
        data = {
            "publicKey": public_key_b64,
            "privateKey": private_key_b64
        }

        # Open the file dialog
        file_path = filedialog.asksaveasfilename(
            parent=root,  
            title="Save Your Keys",  
            defaultextension=".json", 
            initialfile="keys.json",  
            filetypes=[  
                ("JSON files", "*.json"),
                ("All files", "*.*")
            ]
        )

        # If user cancels the dialog, file_path will be empty
        if not file_path:
            return False, None

        # Save the file
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2)

        return True, file_path

    except Exception as e:
        print(f"Error in save_keys_to_json_file: {str(e)}")  # Add error logging
        return False, None

    finally:
        # Clean up
        try:
            root.destroy()
        except Exception as e:
            print(f"Error destroying root window: {str(e)}")  # Add error logging
            pass

def encrypt_and_save_key(private_key_b64: str, derived_key: bytes, username: str) -> bool:
    """
    Encrypt the private key with a random key, then encrypt that key with the derived key.
    Save both to files.

    Args:
        private_key_b64 (str): Base64-encoded private key.
        derived_key (bytes): Key derived from the user's password.
        username (str): Username for file naming.

    Returns:
        bool: True if successful, False otherwise.
    """
    #Prepare JSON
    json_data = json.dumps({"privateKey": private_key_b64}).encode('utf-8')

    #Generate random key and nonce
    key = EncryptionHelper.generate_key()
    nonce = EncryptionHelper.generate_nonce()

    #Encrypt private key JSON
    ciphertext = EncryptionHelper.encrypt(json_data, key, nonce)

    #Prepare data: [nonce][ciphertext]
    combined_data = nonce + ciphertext

    #Save encrypted private key file
    file_name = f"{KEYS_PATH}{username}{BINARY_EXTENSION}"
    try:
        with open(file_name, 'wb') as f:
            f.write(combined_data)
    except IOError as e:
        print(f"Error saving encrypted key file: {str(e)}")
        return False

    #Encrypt and save master key
    if not encrypt_and_save_master_key(key, derived_key, username):
        print("Error saving encrypted key file")
        return False

    return True

def encrypt_and_save_master_key(key_to_encrypt: bytes, derived_key: bytes, username: str) -> bool:
    """
    Encrypt the random key with the derived key and save to file.

    Args:
        key_to_encrypt (bytes): The random key to encrypt.
        derived_key (bytes): The key derived from the user's password.
        username (str): Username for file naming.

    Returns:
        bool: True if successful, False otherwise.
    """
    nonce = EncryptionHelper.generate_nonce()
    encrypted_key = EncryptionHelper.encrypt(key_to_encrypt, derived_key, nonce)
    combined_data = nonce + encrypted_key

    file_path = f"{MASTER_KEY_PATH}{username}{BINARY_EXTENSION}"
    try:
        with open(file_path, 'wb') as f:
            f.write(combined_data)
        return True
    except IOError as e:
        print(f"Error saving master key file: {str(e)}")
        return False 
    


def decrypt_and_reencrypt_user_file(username: str, old_password: str, old_salt: str, new_password: str, new_salt: str) -> bool:
    """
    Decrypt a user's encrypted file using their old password and salt, then re-encrypt it with their new password and salt.

    Args:
        username (str): The username of the user
        old_password (str): The user's old password
        old_salt (str): The user's old salt (base64 encoded)
        new_password (str): The user's new password
        new_salt (str): The user's new salt (base64 encoded)

    Returns:
        bool: True if successful, False otherwise
    """
    crypto = EncryptionHelper()

    # Read encrypted file
    file_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), f"{MASTER_KEY_PATH}{username}{BINARY_EXTENSION}")
    try:
        with open(file_path, 'rb') as file:
            encrypted_data = file.read()
    except IOError as e:
        print(f"Failed to open file for reading: {file_path}")
        return False

    # Extract nonce and ciphertext
    nonce_size = crypto_aead_xchacha20poly1305_ietf_NPUBBYTES
    nonce = encrypted_data[-nonce_size:]
    ciphertext = encrypted_data[:-nonce_size]

    # Derive old key
    old_salt_raw = base64.b64decode(old_salt)
    old_key = derive_key_from_password(old_password, old_salt_raw)

    # Decrypt
    try:
        decrypted_data = crypto.decrypt(
            ciphertext,
            old_key,
            nonce,
            None,  # no additional data
            0
        )
    except Exception as e:
        print(f"Decryption failed: {str(e)}")
        return False

    # Derive new key
    new_salt_raw = base64.b64decode(new_salt)
    new_key = derive_key_from_password(new_password, new_salt_raw)

    # Re-encrypt and save
    return encrypt_and_save_master_key(decrypted_data, new_key, username)

def derive_key_from_password(password: str, salt: bytes, key_len: int = 32) -> bytes:
    """
    Derive a key from a password and salt using Argon2id.

    Args:
        password (str): The password to derive the key from.
        salt (bytes): The salt to use for key derivation.
        key_len (int, optional): Desired key length in bytes. Defaults to 32.

    Returns:
        bytes: The derived key.
    """
    opslimit = pwhash.argon2id.OPSLIMIT_INTERACTIVE
    memlimit = pwhash.argon2id.MEMLIMIT_INTERACTIVE
    return pwhash.argon2id.kdf(key_len, password.encode('utf-8'), salt, opslimit=opslimit, memlimit=memlimit)

def generate_salt(length: int = pwhash.argon2id.SALTBYTES) -> str:
    """
    Generate a random salt and return it as a base64-encoded string.

    Args:
        length (int): Length of the salt in bytes. Defaults to pwhash.argon2id.SALTBYTES.

    Returns:
        str: Base64-encoded salt.
    """
    salt = nacl.utils.random(length)
    salt_b64 = base64.b64encode(salt).decode('utf-8')
    return salt_b64

def decode_salt(salt_b64: str) -> bytes:
    """
    Decode a base64-encoded salt string back to bytes.

    Args:
        salt_b64 (str): Base64-encoded salt.

    Returns:
        bytes: The decoded salt.
    """
    return base64.b64decode(salt_b64)

File: src/desktop-client/client-404/crypto/key_utils.cpp
#include "key_utils.h"
#include <qjsondocument.h>
#include <qjsonobject.h>
#include <sodium.h>
#include <QFileDialog>
#include <QFile>
#include <QMessageBox>
#include <QByteArray>
#include <QString>
#include <QCoreApplication>
#include <iostream>
#include <vector>
#include "constants.h"
#include "core/loginsessionmanager.h"
#include "utils/securebufferutils.h"
#include "utils/securevector.h"


using namespace std;

bool generateSodiumKeyPair(QString &publicKeyBase64, QString &privateKeyBase64) {
    // Generate Ed25519 keys instead of Curve25519
    auto ed25519_pk = make_secure_buffer<crypto_sign_PUBLICKEYBYTES>();
    auto ed25519_sk = make_secure_buffer<crypto_sign_SECRETKEYBYTES>();

    if (crypto_sign_keypair(ed25519_pk.get(), ed25519_sk.get()) != 0) {
        qWarning() << "Failed to generate Ed25519 keypair";
        return false;
    }

    QByteArray pubKeyArray(reinterpret_cast<char*>(ed25519_pk.get()), crypto_sign_PUBLICKEYBYTES);
    QByteArray privKeyArray(reinterpret_cast<char*>(ed25519_sk.get()), crypto_sign_SECRETKEYBYTES);

    publicKeyBase64 = pubKeyArray.toBase64();
    privateKeyBase64 = privKeyArray.toBase64();

    return true;
}


bool encryptAndSaveKey(QWidget *parent, const QString &privateKey, const unsigned char *derivedKey, QString username, bool (*saveFuncPtrArg)(const QString&, const SecureVector&) = saveFile) {
    QJsonObject json;
    json["privateKey"] = privateKey;

    QJsonDocument doc(json);
    QByteArray jsonData = doc.toJson();

    shared_ptr<EncryptionHelper> crypto = make_shared<EncryptionHelper>();

    AnotherBasePage crypto2(*crypto);

    auto key = make_secure_buffer<crypto_aead_xchacha20poly1305_ietf_KEYBYTES>();
    auto nonce = make_secure_buffer<crypto_aead_xchacha20poly1305_ietf_NPUBBYTES>();

    // Generate key and nonce
    crypto->generateKey(key.get(), crypto_aead_xchacha20poly1305_ietf_KEYBYTES);
    crypto->generateNonce(nonce.get(), crypto_aead_xchacha20poly1305_ietf_NPUBBYTES);

    //Encrypt private key file
    SecureVector ciphertext;
    bool success = false;
    try {
        ciphertext = encryptData(jsonData, key.get(), nonce.get(), crypto);
    } catch (const exception &e) {
        QMessageBox::critical(parent, "Encryption Error", e.what());
    }

    // Prepare data: [nonce][ciphertext]
    SecureVector combinedData(crypto_aead_xchacha20poly1305_ietf_NPUBBYTES + ciphertext.size());
    std::copy(nonce.get(), nonce.get() + crypto_aead_xchacha20poly1305_ietf_NPUBBYTES, combinedData.begin());    // Copies the nonce to the start of the buffer
    std::copy(ciphertext.data(), ciphertext.data() + ciphertext.size(),  combinedData.begin() + crypto_aead_xchacha20poly1305_ietf_NPUBBYTES); // Copy the ciphertext to the buffer right after the nonce

    //Save encrypted private key file
    QString fileName = QCoreApplication::applicationDirPath() + keysPath + username + binaryExtension; //encryptedKey_username.bin
    bool (*saveFuncPtr)(const QString&, const SecureVector&) = saveFuncPtrArg; //function pointer
    if (!saveFuncPtr(fileName, combinedData)) {
        cout << "Error saving file" << endl;
        jsonData.fill(0);
        jsonData.clear();
        return false;
    }

    //clear data
    jsonData.fill(0);
    jsonData.clear();


    //Encrypt and save master key
    if(!encryptAndSaveMasterKey(key.get(), crypto_aead_xchacha20poly1305_ietf_KEYBYTES, derivedKey, crypto, username)){
        cout << "Error saving encrypted key file" << endl;
        return false;
    }

    LoginSessionManager::getInstance().setSession(username, key.get(), crypto_aead_xchacha20poly1305_ietf_KEYBYTES);
    return true;
}

bool encryptAndSaveMasterKey(const unsigned char *keyToEncrypt, size_t keyLen, const unsigned char *derivedKey, shared_ptr<EncryptionHelper> crypto, QString username)
{
    // Generate nonce
    auto nonce = make_secure_buffer<crypto_aead_xchacha20poly1305_ietf_NPUBBYTES>();
    crypto->generateNonce(nonce.get(), crypto_aead_xchacha20poly1305_ietf_NPUBBYTES);


    // Encrypt the key
    SecureVector encryptedKey = crypto->encrypt(
        keyToEncrypt,
        keyLen,
        derivedKey,
        nonce.get(),
        nullptr,
        0
        );

    // Prepare data: [nonce][encryptedKey]
    SecureVector combinedData(crypto_aead_xchacha20poly1305_ietf_NPUBBYTES + encryptedKey.size());
    std::copy(nonce.get(), nonce.get() + crypto_aead_xchacha20poly1305_ietf_NPUBBYTES, combinedData.begin());    // Copies the nonce to the start of the buffer
    std::copy(encryptedKey.data(), encryptedKey.data() + encryptedKey.size(), combinedData.begin() + crypto_aead_xchacha20poly1305_ietf_NPUBBYTES);  // Copy the ciphertext to the buffer right after the nonce

    // Save to file
    QString filePath = QCoreApplication::applicationDirPath() + masterKeyPath + username + binaryExtension;//masterKey.bin;
    bool (*saveFuncPtr)(const QString&, const SecureVector&) = saveFile; //function pointer
    bool success = saveFuncPtr(filePath, combinedData);
    return success;
}


SecureVector encryptData(const QByteArray &plaintext, unsigned char *key, unsigned char *nonce, shared_ptr<EncryptionHelper> crypto){

    const unsigned char* plaintext_ptr;
    unsigned long long plaintext_len;
    try {
        plaintext_ptr = reinterpret_cast<const unsigned char*>(plaintext.constData());
        plaintext_len = static_cast<unsigned long long>(plaintext.size());
    } catch (const exception& e) {
        qWarning() << "Exception:" << e.what();
    }

    // Encrypt with no metadata
    return crypto->encrypt(
        plaintext_ptr,
        plaintext_len,
        key,
        nonce,
        nullptr,
        0
        );
}

bool decryptAndReencryptUserFile(const QString& username, const QString& oldPassword, const QString& oldSalt, const QString& newPassword, const QString& newSalt)
{
    shared_ptr<EncryptionHelper> crypto = make_shared<EncryptionHelper>();

    //Read encrypted file
    QString filePath = QCoreApplication::applicationDirPath() + masterKeyPath + username + binaryExtension;
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        qDebug() << "Failed to open file for reading:" << filePath;
        return false;
    }
    QByteArray encryptedData = file.readAll();
    file.close();

    //Extract nonce and ciphertext
    const int nonceSize = crypto_aead_xchacha20poly1305_ietf_NPUBBYTES;
    QByteArray nonce = encryptedData.left(nonceSize);  // Get nonce from start
    QByteArray ciphertext = encryptedData.mid(nonceSize);  // Get ciphertext after nonce

    //Derive old key
    QByteArray oldSaltRaw = QByteArray::fromBase64(oldSalt.toUtf8());
    unsigned char oldKey[crypto_aead_xchacha20poly1305_ietf_KEYBYTES];
    deriveKeyFromPassword(oldPassword.toStdString(), reinterpret_cast<const unsigned char*>(oldSaltRaw.constData()), oldKey, sizeof(oldKey));

    //Decrypt
    SecureVector decryptedData;
    try {
        decryptedData = crypto->decrypt(
            reinterpret_cast<const unsigned char*>(ciphertext.constData()), ciphertext.size(),
            oldKey,
            reinterpret_cast<const unsigned char*>(nonce.constData()),
            nullptr, // no additional data
            0
            );
    } catch (const std::exception& e) {
        qDebug() << "Decryption failed:" << e.what();
        return false;
    }

    //Derive new key
    QByteArray newSaltRaw = QByteArray::fromBase64(newSalt.toUtf8());
    unsigned char newKey[crypto_aead_xchacha20poly1305_ietf_KEYBYTES];
    deriveKeyFromPassword(newPassword.toStdString(), reinterpret_cast<const unsigned char*>(newSaltRaw.constData()), newKey, sizeof(newKey));

    return encryptAndSaveMasterKey(decryptedData.data(), decryptedData.size(), newKey, crypto, username);



}

bool decryptMasterKey(const QString& username, const QString& password, const QString& salt) {
    shared_ptr<EncryptionHelper> crypto = make_shared<EncryptionHelper>();

    // Read encrypted file
    QString filePath = QCoreApplication::applicationDirPath() + masterKeyPath + username + binaryExtension;
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        qDebug() << "Failed to open file for reading:" << filePath;
        return false;
    }
    QByteArray encryptedData = file.readAll();
    file.close();

    // Extract nonce and ciphertext
    const int nonceSize = crypto_aead_xchacha20poly1305_ietf_NPUBBYTES;
    QByteArray nonce = encryptedData.left(nonceSize);  // Get nonce from start
    QByteArray ciphertext = encryptedData.mid(nonceSize);  // Get ciphertext after nonce

    // Derive key from password and salt
    QByteArray saltRaw = QByteArray::fromBase64(salt.toUtf8());
    unsigned char derivedKey[crypto_aead_xchacha20poly1305_ietf_KEYBYTES];
    deriveKeyFromPassword(password.toStdString(), reinterpret_cast<const unsigned char*>(saltRaw.constData()), derivedKey, sizeof(derivedKey));

    // Decrypt
    SecureVector decryptedData;
    try {
        decryptedData = crypto->decrypt(
            reinterpret_cast<const unsigned char*>(ciphertext.constData()),
            ciphertext.size(),
            derivedKey,
            reinterpret_cast<const unsigned char*>(nonce.constData()),
            nullptr, // no additional data
            0
            );
    } catch (const std::exception& e) {
        qDebug() << "Decryption failed:" << e.what();
        return false;
    }

    // Set the session with the decrypted master key
    LoginSessionManager::getInstance().setSession(username, decryptedData.data(), decryptedData.size());
    return true;
}

//Function overloading
bool saveFile(const QString &filePath, const SecureVector &data) {
    QFile file(filePath);
    if (file.open(QIODevice::WriteOnly)) {
        file.write(reinterpret_cast<const char*>(data.data()), static_cast<qint64>(data.size()));
        file.close();
        return true;
    }
    return false;
}

bool saveFile(QWidget *parent, const QJsonObject &json, const QString &defaultName) {
    QJsonDocument doc(json);
    QByteArray jsonData = doc.toJson();

    QString fileName = QFileDialog::getSaveFileName(parent, "Save Keys", defaultName, "JSON Files (*.json);;All Files (*)");
    if (!fileName.isEmpty()) {
        QFile file(fileName);
        if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
            file.write(jsonData);
            file.close();
            return true;
        } else {
            QMessageBox::warning(parent, "Error", "Failed to open file for writing.");
        }
    }
    return false;
}

bool (*getSaveFileFunction())(const QString&, const SecureVector&) {
    // Declare the function pointer
    bool (*saveFunc)(const QString&, const SecureVector&) = &saveFile;

    // Return the function pointer
    return saveFunc;
}


